packages/block-editor/src/hooks/block-style-variation.js

```diff
@@ -0,0 +1,136 @@
+/**
+ * WordPress dependencies
+ */
+import { getBlockTypes, store as blocksStore } from '@wordpress/blocks';
+import { useSelect } from '@wordpress/data';
+import { useContext, useMemo } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import {
+	GlobalStylesContext,
+	toStyles,
+	getBlockSelectors,
+} from '../components/global-styles';
+import { useStyleOverride } from './utils';
+import { store as blockEditorStore } from '../store';
+
+function getVariationNameFromClass( className ) {
+	const match = className?.match( /\bis-style-(?!default)(\S+)\b/ );
+	return match ? match[ 1 ] : null;
+}
+
+function useBlockSyleVariation( name, variation, clientId ) {
+	// Prefer global styles data in GlobalStylesContext, which are available
+	// if in the site editor. Otherwise fall back to whatever is in the
+	// editor settings and available in the post editor.
+	// This can be updated once the global styles data is consistently
+	// available across the editors.
+	const { merged: mergedConfig } = useContext( GlobalStylesContext );
+	const { globalSettings, globalStyles } = useSelect( ( select ) => {
+		const { __experimentalFeatures, __experimentalStyles } =
+			select( blockEditorStore ).getSettings();
+		return {
+			globalSettings: __experimentalFeatures,
+			globalStyles: __experimentalStyles,
+		};
+	}, [] );
+
+	return useMemo( () => {
+		const styles = mergedConfig?.styles ?? globalStyles;
+		const variationStyles =
+			styles?.blocks?.[ name ]?.variations?.[ variation ];
+
+		return {
+			settings: mergedConfig?.settings ?? globalSettings,
+			// The variation style data is all that is needed to generate
+			// the styles for the current application to a block. The variation
+			// name is updated to match the instance specific class name.
+			styles: {
+				blocks: {
+					[ name ]: {
+						variations: {
+							[ `${ variation }-${ clientId }` ]: variationStyles,
+						},
+					},
+				},
+			},
+		};
+	}, [
+		mergedConfig,
+		globalSettings,
+		globalStyles,
+		variation,
+		clientId,
+		name,
+	] );
+}
+
+// Rather than leveraging `useInstanceId` here, the `clientId` is used.
+// This is so that the variation style override's ID is predictable
+// when the order of applied style variations changes.
+function useBlockProps( { name, className, clientId } ) {
+	const variation = getVariationNameFromClass( className );
+	const variationClass = `is-style-${ variation }-${ clientId }`;
+
+	const { getBlockStyles } = useSelect( blocksStore );
+
+	const { settings, styles } = useBlockSyleVariation(
+		name,
+		variation,
+		clientId
+	);
```

Comment:
There's probably a performance degradation because this adds a block editor subscription for every block with a className?

Reply:
At one point I had the variations block support add a `style.variation` block attribute. I could revisit that approach.

Currently, a block style variation is only set via an `is-style-*` class, as it is on trunk. If a variation attribute is set it would also allow the feature to skip having to regex out variation names before retrieving data etc.